Library/Homebrew/migrator.rb

```diff
@@ -284,18 +291,27 @@ def backup_oldname
     end
 
     if oldkeg_linked?
-      begin
-        # The keg used to be linked  and when we backup everything we restore
-        # Cellar/oldname, the target also gets restored, so we are able to
-        # create a keg using its old path
-        keg = Keg.new(Pathname.new(oldkeg.to_s))
-        keg.link
-      rescue Keg::LinkError
-        keg.unlink
-        raise
-      rescue Keg::AlreadyLinkedError
-        keg.unlink
-        retry
+      # The keg used to be linked and when we backup everything we restore
+      # Cellar/oldname, the target also gets restored, so we are able to
+      # create a keg using its old path
+      if old_linked_keg_record
+        begin
+          oldkeg.link
+        rescue Keg::LinkError
+          oldkeg.unlink
+          raise
+        rescue Keg::AlreadyLinkedError
+          oldkeg.unlink
+          retry
+        end
+      else
+        begin
+          oldkeg.optlink
+        rescue Keg::LinkError => e
+          onoe "Failed to create #{formula.opt_prefix}"
```

Comment:
Final question before I merge this: why do you need to rescue here. Also `oldkeg.optlink` failure means failing to create old opt prefix rather than `formula.opt_prefix`.

I suggest you just make it:
```ruby
else
  oldkeg.optlink
end
```